src/System.Collections/src/System/Collections/Generic/SortedSet.cs

```diff
@@ -92,76 +92,40 @@ public SortedSet(IEnumerable<T> collection, IComparer<T> comparer)
-                    return;
-                }
-
-                // pre order way to replicate nodes
-                Stack<Node> theirStack = new Stack<SortedSet<T>.Node>(2 * Log2(baseSortedSet.Count) + 2);
-                Stack<Node> myStack = new Stack<SortedSet<T>.Node>(2 * Log2(baseSortedSet.Count) + 2);
-                Node theirCurrent = baseSortedSet._root;
-                Node myCurrent = (theirCurrent != null ? new SortedSet<T>.Node(theirCurrent.Item, theirCurrent.IsRed) : null);
-                _root = myCurrent;
-                while (theirCurrent != null)
-                {
-                    theirStack.Push(theirCurrent);
-                    myStack.Push(myCurrent);
-                    myCurrent.Left = (theirCurrent.Left != null ? new SortedSet<T>.Node(theirCurrent.Left.Item, theirCurrent.Left.IsRed) : null);
-                    theirCurrent = theirCurrent.Left;
-                    myCurrent = myCurrent.Left;
-                }
-                while (theirStack.Count != 0)
+            // These are explicit type checks in the mould of HashSet. It would have worked better with
+            // something like an ISorted<T> interface. (We could make this work for SortedList.Keys, etc.)
+            SortedSet<T> sortedSet = collection as SortedSet<T>;
+            if (sortedSet != null && !(sortedSet is TreeSubSet) && HasEqualComparer(sortedSet))
+            {
+                if (sortedSet.Count > 0)
                 {
-                    theirCurrent = theirStack.Pop();
-                    myCurrent = myStack.Pop();
-                    Node theirRight = theirCurrent.Right;
-                    Node myRight = null;
-                    if (theirRight != null)
-                    {
-                        myRight = new SortedSet<T>.Node(theirRight.Item, theirRight.IsRed);
-                    }
-                    myCurrent.Right = myRight;
-
-                    while (theirRight != null)
-                    {
-                        theirStack.Push(theirRight);
-                        myStack.Push(myRight);
-                        myRight.Left = (theirRight.Left != null ? new SortedSet<T>.Node(theirRight.Left.Item, theirRight.Left.IsRed) : null);
-                        theirRight = theirRight.Left;
-                        myRight = myRight.Left;
-                    }
+                    Debug.Assert(sortedSet._root != null);
+                    _count = sortedSet._count;
+                    _root = sortedSet._root.DeepClone(_count);
                 }
-                _count = baseSortedSet._count;
+                return;
             }
-            else
+            
+            int count;
+            T[] elements = EnumerableHelpers.ToArray(collection, out count);
+            if (count > 0)
             {
-                int count;
-                T[] els = EnumerableHelpers.ToArray(collection, out count);
-                if (count > 0)
+                Array.Sort(elements, 0, count, comparer);
```

Comment:
Why did you remove setting comparer to _comparer?  If comparer is null, aren't the below calls to comparer.Compare going to blow up?